File: 498-Diagonal Traverse.py
class Solution(object):
    def findDiagonalOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix or not matrix[0]:
            return []
        
        M = len(matrix[0])
        N = len(matrix)
        res = []
        intermediate = []
        
        for row in range(N+M-1):
            intermediate = []
            if row < M:
                r = 0
            else:
                r = row - M + 1
            
            if row < M:
                c = row
            else:
                c = M-1
                
            while r < N and c >= 0:
                intermediate.append(matrix[r][c])
                r += 1
                c-= 1
            if row%2 == 1:
                res+=intermediate
            else:
                res+=intermediate[::-1]
            
        return res
                

File: C++/498-Diagonal Traverse.cpp
// First AC
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        bool reverse = false;
        int row = mat.size(), col = mat[0].size();
        vector<int> res;

        for (int sum = 0; sum <= row+col-2; sum++) {
            if (reverse) {
                for (int c=min(sum, col-1); sum-c<=row-1 && c >=0; c--) {
                    int r = sum-c;
                    res.push_back(mat[r][c]);
                }
            } else {
                for (int r=min(sum, row-1); sum-r<=col-1 && r>=0; r--) {
                    int c = sum-r;
                    res.push_back(mat[r][c]);
                }
            }
            reverse = !reverse;
        }
        
        return res;
    }
};


// O(row*col)
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        int row = mat.size(), col = mat[0].size();
        int r=0, c=0;
        vector<int> res;

        while (res.size() < row*col) {
            res.push_back(mat[r][c]);

            if ((r+c)%2 == 0) {
                if (c == col-1) {
                    r++;
                } else if (r == 0) {
                    c++;
                } else {
                    c++;
                    r--;
                }
            } else {
                if (r == row-1) {
                    c++;
                } else if (c == 0) {
                    r++;
                } else {
                    c--;
                    r++;
                }                
            }
        }

        return res;
    }
};